pkg/resultstore/store.go

```diff
@@ -0,0 +1,189 @@
+package resultstore
+
+import (
+	"context"
+	"log"
+	"sort"
+	"sync"
+	"time"
+
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+)
+
+// TODO: make these configurable?
+var (
+	notifyQBufferSize = 128
+	storeTTL          = time.Minute
+	gcInterval        = time.Minute * 5
+)
+
+// element is an internal representation of a result.
+type element struct {
+	data    ocr2keepers.CheckResult
+	addedAt time.Time
+}
+
+// resultStore implements ResultStore.
+type resultStore struct {
+	lggr *log.Logger
+
+	data map[string]element
+	lock sync.RWMutex
+
+	notifications chan ocr2keepers.Notification
+}
+
+func New(lggr *log.Logger) *resultStore {
+	return &resultStore{
+		lggr:          lggr,
+		data:          make(map[string]element),
+		lock:          sync.RWMutex{},
+		notifications: make(chan ocr2keepers.Notification, notifyQBufferSize),
+	}
+}
+
+// Start starts the store, it spins up a goroutine that runs the garbage collector every gcInterval.
+func (s *resultStore) Start(pctx context.Context) error {
+	s.lggr.Println("starting result store")
+
+	go func() {
+		ctx, cancel := context.WithCancel(pctx)
+		defer cancel()
+
+		ticker := time.NewTicker(gcInterval)
+		for {
+			select {
+			case <-ticker.C:
+				s.gc()
+			case <-ctx.Done():
+				s.lggr.Println("result store context done, stopping gc")
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Notifications returns a channel that can be used to receive notifications about evicted/removed items in the store.
+func (s *resultStore) Notifications() <-chan ocr2keepers.Notification {
+	return s.notifications
+}
+
+// Add adds element/s to the store.
+func (s *resultStore) Add(results ...ocr2keepers.CheckResult) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	added := 0
+	for _, result := range results {
+		id := result.Payload.ID
+		_, ok := s.data[id]
+		if ok {
+			// if the element is already exists, we do noting
+			continue
+		}
+		added++
+		s.data[id] = element{data: result, addedAt: time.Now()}
+	}
+}
+
+// Remove removes element/s from the store.
+func (s *resultStore) Remove(ids ...string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	for _, id := range ids {
+		s.remove(id)
+	}
+}
+
+// View returns a copy of the data in the store.
+// It accepts filters that can be used to prepare the results view.
+// NOTE: we apply filters while holding the read lock, these functions must not block.
+func (s *resultStore) View(opts ...ocr2keepers.ViewOpt) ([]ocr2keepers.CheckResult, error) {
+	filters, comparators, limit := ocr2keepers.ViewOpts(opts).Apply()
+
+	var results []ocr2keepers.CheckResult
+	s.lock.RLock()
```

Comment:
Does it make sense to extract lines 108-132 into a function? That way you can lock and defer the unlock for just this section of code, rather than having to specify where the unlocking should happen?